lib/Admin.php

```diff
@@ -0,0 +1,226 @@
+	 * @var    Plugin $plugin This plugin's instance.
+	 */
+	private $plugin;
+
+	/**
+	 * Initialize the class and set its properties.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param Plugin $plugin This plugin's instance.
+	 */
+	public function __construct( Plugin $plugin ) {
+		$this->plugin = $plugin;
+	}
+
+	/**
+	 * The primary handler for user authentication.
+	 *
+	 * @since     0.1.0
+	 * @param     mixed    $user    The current user (or bool) passing through the filter.
+	 * @return    mixed             A user on success, or false on failure.
+	 */
+	public function auth_handler( $user ) {
+
+		// Don't authenticate twice
+		if ( ! empty( $user ) ) {
+			return $user;
+		}
+
+		// Check for request headers
+		if ( ! $this->check_request_headers() ) {
+			return $user;
+		}
+
+		// Fetch user ID
+		$user_id = $this->find_user_id_by_key( $_SERVER[ 'HTTP_X_API_KEY' ] );
+
+		if ( ! $user_id ) {
+			$this->set_response_headers( 'FAIL api key' );
+			return false;
+		}
+
+		// Check timestamp
+		if ( ! $this->valid_timestamp() ) {
+			$this->set_response_headers( 'FAIL timestamp' );
+			return false;
+		}
+
+		// Check for the proper HTTP Parameters
+		$signature_args = array(
+			'api_key'        => $_SERVER[ 'HTTP_X_API_KEY' ],
+			'ip'             => $_SERVER[ 'REMOTE_ADDR' ],
+			'request_method' => $_SERVER[ 'REQUEST_METHOD' ],
+			'request_post'   => $_POST,
+			'request_uri'    => $_SERVER[ 'REQUEST_URI' ],
+			'timestamp'      => $_SERVER[ 'HTTP_X_API_TIMESTAMP' ],
+		);
+
+		$user_secret   = \get_user_meta( $user_id, 'rest_api_secret', true );
+		$signature_gen = self::generate_signature( $signature_args, $user_secret );
+		$signature     = $_SERVER[ 'HTTP_X_API_SIGNATURE' ];
+
+		if ( $signature_gen !== $signature ) {
+			$this->set_response_headers( 'FAIL signature' );
+			return false;
+		}
+
+		// Set headers
+		$this->set_response_headers( 'OK ' . $user_id );
+
+		return $user_id;
+	}
+
+	/**
+	 * Checks for the presence of required request headers.
+	 *
+	 * @access    private
+	 * @since     1.0.0
+	 * @return    bool    True if the request headers are present, false otherwise.
+	 */
+	private function check_request_headers() {
+		return
+			isset( $_SERVER[ 'HTTP_X_API_KEY' ] ) &&
+			isset( $_SERVER[ 'HTTP_X_API_TIMESTAMP' ] ) &&
+			isset( $_SERVER[ 'HTTP_X_API_SIGNATURE' ] );
+	}
+
+	/**
+	 * Fetches a user ID by API key.
+	 *
+	 * @access    private
+	 * @since     0.1.0
+	 * @param     string    $api_key    The API key attached to a user.
+	 * @return    mixed                 A user ID on success, or false on failure.
+	 */
+	private function find_user_id_by_key( $api_key ) {
+
+		$user_args = array(
+			'meta_query' => array(
+				array(
+					'key'   => 'rest_api_key',
+					'value' => $api_key,
+				),
+			),
+			'number'     => 1,
+			'fields'     => array( 'ID' ),
+		);
+
+		$user = \get_users( $user_args );
+
+		if ( empty( $user ) || ! is_array( $user ) ) {
+			return false;
+		}
+
+		return $user[0]->ID;
```

Comment:
Should we not check for the fact that there is more than one user with the same API key?

Probably it's done while adding the API key, but if we get a response here that is more than 1 in length, we would be catching a previous error in the API key attribution. I would suggest checking if the length is bigger than 1 and, for now, return false (trusting that this might not happen because we attribute a unique API key per user).